audioipc/src/shm.rs

```diff
@@ -4,263 +4,286 @@
+            self.view.get_slice(size)
         }
-        // TODO: Track how much is in the shm area.
-        if buf.len() <= self.mmap.len() {
-            self.mmap[..buf.len()].copy_from_slice(buf);
-            atomic::fence(atomic::Ordering::Release);
-            Ok(())
-        } else {
-            bail!("mmap size");
+
+        pub unsafe fn get_mut_slice(&mut self, size: usize) -> Result<&mut [u8]> {
+            self.view.get_mut_slice(size)
         }
     }
 }
 
-pub struct SharedMemMutSlice {
-    mmap: Arc<UnsafeCell<MmapMut>>,
-}
+#[cfg(windows)]
+mod windows {
+    use super::*;
+    use std::ptr;
+    use winapi::{
+        shared::ntdef::HANDLE,
+        um::{
+            handleapi::CloseHandle,
+            memoryapi::{MapViewOfFile, UnmapViewOfFile, FILE_MAP_ALL_ACCESS},
+            winbase::CreateFileMappingA,
+            winnt::PAGE_READWRITE,
+        },
+    };
+
+    use crate::INVALID_HANDLE_VALUE;
+
+    pub struct SharedMem {
+        handle: HANDLE,
+        view: SharedMemView,
+    }
 
-impl SharedMemMutSlice {
-    pub fn from(file: &File, size: usize) -> Result<SharedMemMutSlice> {
-        let mmap = unsafe { MmapOptions::new().map_mut(file)? };
-        assert_eq!(mmap.len(), size);
-        let mmap = Arc::new(UnsafeCell::new(mmap));
-        Ok(SharedMemMutSlice { mmap })
+    unsafe impl Send for SharedMem {}
+
+    impl Drop for SharedMem {
+        fn drop(&mut self) {
+            unsafe {
+                let ok = UnmapViewOfFile(self.view.view);
+                assert_ne!(ok, 0);
+                if self.handle != INVALID_HANDLE_VALUE {
+                    let ok = CloseHandle(self.handle);
+                    assert_ne!(ok, 0);
+                }
+            }
+        }
     }
 
-    pub fn get_mut_slice(&mut self, size: usize) -> Result<&mut [u8]> {
-        if size == 0 {
-            return Ok(&mut []);
+    impl SharedMem {
+        pub fn new(_id: &str, size: usize) -> Result<(SharedMem, PlatformHandle)> {
+            unsafe {
+                let handle = CreateFileMappingA(
+                    INVALID_HANDLE_VALUE,
+                    ptr::null_mut(),
+                    PAGE_READWRITE,
+                    0,
+                    size.try_into().unwrap(),
+                    ptr::null(),
+                );
+                if handle.is_null() {
+                    return Err(std::io::Error::last_os_error().into());
+                }
+
+                let view = MapViewOfFile(handle, FILE_MAP_ALL_ACCESS, 0, 0, size);
+                if view.is_null() {
+                    return Err(std::io::Error::last_os_error().into());
+                }
+
+                let handle2 = PlatformHandle::duplicate(handle)?;
+                Ok((
+                    SharedMem {
+                        handle,
+                        view: SharedMemView { view, size },
+                    },
+                    handle2,
```

Comment:
Is it possible to omit the duplicate `handle` here? or replace it with another a different type? What is the main purpose to use `handle2`?

Reply:
`handle2` is sent from the AudioIPC server to the client as the reference used to instantiate the `SharedMem` on the client side via `SharedMem::from`.

Reply:
I could split it into `SharedMem::new(...) -> Result<SharedMem>` and `SharedMem::get_handle(&self) -> Result<PlatformHandle>` if you think that'd be clearer.

Reply:
I don't have a strong opinon. I am ok if there is some comment for it.

Reply:
I'll take care of this in a separate change.